std/file.d

```diff
@@ -5119,3 +5119,86 @@ string tempDir() @trusted
     myFile.write("hello");
     assert(myFile.readText == "hello");
 }
+
+/**
+Returns the available space in a file system.
+
+On `Windows`, this function uses the result of the Windows API function
+$(LINK2 https://msdn.microsoft.com/en-us/library/windows/desktop/aa364937(v=vs.85).aspx, $(D GetDiskFreeSpaceExW)).
+
+On `POSIX` platforms, it uses the $(LINK2 http://man7.org/linux/man-pages/man3/statvfs.3.html,
+$(D statvfs)) function. On `FreeBSD` however, it uses the $(LINK2 https://www.freebsd.org/cgi/man.cgi?query=statfs&sektion=2,
+$(D statfs)) function.
+
+Params:
+    path = a valid path to any file in the file system
+Returns:
+    Available space in bytes
+
+Throws:
+$(LREF FileException) in case of failure
+*/
+ulong getAvailableDiskSpace(string path) @safe
+{
+    import std.exception : enforce;
+    import std.conv : text;
+
+    version (Windows)
+    {
+        import core.sys.windows.winbase : GetDiskFreeSpaceExW, GetLastError;
+        import core.sys.windows.winnt : ULARGE_INTEGER;
+        import std.internal.cstring : tempCStringW;
+
+        ULARGE_INTEGER freeBytesAvailable;
+        auto err = () @trusted {
+            return GetDiskFreeSpaceExW(path.tempCStringW(), &freeBytesAvailable, null, null);
+        } ();
+        enforce(err != 0,
+            new FileException(text("Cannot get available disk space: ", GetLastError())));
+
+        return freeBytesAvailable.QuadPart;
+    }
+    else version (Posix)
+    {
+        import std.string : toStringz;
+
+        version (FreeBSD)
+        {
+            import core.sys.posix.sys.statvfs : statfs, statfs_t;
+
+            statfs_t stats = void;
+            auto err = () @trusted {
+                return statfs(path.toStringz(), &stats);
+            } ();
+            enforce(err == 0,
+                new FileException(text("Cannot get available disk space: ", err)));
+
+            return stats.f_bavail * stats.f_bsize;
+        }
+        else
+        {
+            import core.sys.posix.sys.statvfs : statvfs, statvfs_t;
+
+            statvfs_t stats = void;
+            auto err = () @trusted {
+                return statvfs(path.toStringz(), &stats);
+            } ();
+            enforce(err == 0,
+                new FileException(text("Cannot get available disk space: ", err)));
+
+            return stats.f_bavail * stats.f_frsize;
+        }
+    }
+    else static assert(0, "Unsupported platform");
+}
+
+///
+@safe unittest
+{
+    import std.exception : assertThrown;
+
+    auto space = getAvailableDiskSpace(".");
+    assert(space > 0);
+
+    assertThrown!FileException(getAvailableDiskSpace("ThisFileDoesNotExist123123"));
+}
```

Comment:
I think we can do more tests here. Simple create a directory with `deleteme` and write files to it and check the resulting file size (if this varies on Windows, `version(Posix)` it).

Example:

https://dlang.org/phobos-prerelease/std_file.html#mkdir

Reply:
That'd be a bit excessive - we'd be testing the OS primitives.

Reply:
You should test that the function works for `null` and `""` (empty strings) input. See e.g. https://issues.dlang.org/show_bug.cgi?id=14980